kubeshop/botkube
# Add thread mate plugin

## Description

Changes proposed in this pull request:

- Add thread mate plugin
  - randomize initial next idx value
  - opt to define ack message emoji
  - opt to bind for channel messages
  - persist processed activity

## Testing

1. Build plugin: `PLUGIN_TARGETS="thread-mate" make build-plugins-single`
2. Create config:
    <details><summary>Details</summary>
    <p>

    ```yaml
    communications:
      default-group:
        socketSlack:
          enabled: true
          channels:
            general:
              name: backoffice
              bindings:
                executors:
                  - helping-hands-assigner
                sources: []
            default:
              name: helping-hands
              textMessageBindings:
                event: ChannelMessage # currently only ChannelMessage is supported. Other options that we may add in the future: ChannelMessage, ThreadMessage, AllMessages
                messages:
                  - text:
                      include: ".*" # all messages
                    users:
                      exclude: ["U02KKBR5PE1:Mateusz", "etc"]
                    processedEmojiIndicator: "robot_face" # empty - disabled
                    command: "thread-mate pick --bk-cmd-header 'Help assistance' "
                    executors:
                      - helping-hands-assigner
              bindings:
                executors: []
                sources: []
    
          appToken: "xapp-"
          botToken: "xoxb-"
    
    executors:
      helping-hands-assigner:
        botkube/thread-mate:
          enabled: true
          context:
            rbac:
              group:
                type: Static
                static:
                  values: [ "system:masters" ]
          config:
            pick:
              # userCooldownTime represents the time that must elapse for a given user's message before the "pick" method can be used again.
              userCooldownTime: 3m
              #messageTemplate: |
                # ...
            roundRobin:
              assignees: ["U02KKBR5PE1:Mateusz", "etc"]
              groupName: "botkube"
    
    aliases:
      act-all:
        command: thread-mate get activity
        displayName: "Get thread activity"
      act:
        command: thread-mate get activity --assignee-ids
        displayName: "Get thread activity"
    
    plugins:
      cacheDir: "/tmp/plugins"
      repositories:
        botkube:
          url: http://host.k3d.internal:3010/botkube.yaml
    
    settings:
      log:
        level: "debug"
        formatter: text
      clusterName: "labs"
      upgradeNotifier: false
    
    analytics:
      disable: true
    ```
    
    </p>
    </details> 

3. Start plugin server: `env PLUGIN_SERVER_HOST=http://host.k3d.internal go run test/helpers/plugin_server.go`
4. Run Botkube: `env LOG_LEVEL_EXECUTOR_BOTKUBE_THREAD_MATE=debug BOTKUBE_CONFIG_PATHS="./helm/botkube/values.yaml,/tmp/mate-cfg.yaml" go run ./cmd/botkube-agent/main.go`
5.  Test:
    - round-robin rotation
    - getting DMs
    - use custom emoji
    - cooldown window for using `pick`
        - if a user sends e.g. 3 messages on main thread within X minutes, only the first one is used to select a support person. 
    - `@Botkube thread-mate get activity`:
        - takeover button
        - mark as resolved
        - export data as markdown and csv
        - search options
    - define different message template
    - Botkube should ignore messages from the support person
    - Thread messages should be ignored
    - restart Botkube, activity should be preserved

## Related issue(s)

- https://github.com/kubeshop/botkube-cloud/issues/631

## Notes
Make sure that your socket slack app is subscribed to channel events
<img width="910" alt="Screenshot 2023-10-11 at 22 55 24" src="https://github.com/kubeshop/botkube/assets/17568639/7f982c2a-c3ca-4840-9f15-7dae35b79b7e">
